lib/net/http/exceptions.rb

```diff
@@ -4,12 +4,37 @@ module Net
   # You cannot use Net::HTTPExceptions directly; instead, you must use
   # its subclasses.
   module HTTPExceptions
+    # Valid keys for pattern matching via #deconstruct_keys.
+    PATTERN_MATCHING_KEYS = %i[message response].freeze
+
     def initialize(msg, res)   #:nodoc:
       super msg
       @response = res
     end
     attr_reader :response
     alias data response    #:nodoc: obsolete
+
+    # Returns a hash of exception attributes for pattern matching.
+    #
+    # Valid keys are: +:message+, +:response+
+    #
+    # Example:
+    #
+    #   begin
+    #     http.request(req)
+    #   rescue => e
+    #     case e
+    #     in HTTPRetriableError[response: { code: '503' }]
```

Comment:
Assuming that https://github.com/ruby/net-http/pull/240 is merged this is possible, otherwise nested deconstruction is not _quite_ possible hence the deviation in the tests.